Disco_Tie/code.py

```diff
@@ -0,0 +1,237 @@
+        * (output_max - output_min)
+
+# Remove DC bias before computing RMS.
+def normalized_rms(values):
+    minbuf = int(mean(values))
+    samples_sum = sum(
+        float(sample - minbuf) * (sample - minbuf)
+        for sample in values
+    )
+
+    return math.sqrt(samples_sum / len(values))
+
+def mean(values):
+    return sum(values) / len(values)
+
+def volume_color(volume):
+    return 200, volume * (255 // NUM_PIXELS), 0
+
+# Set up NeoPixels and turn them all off.
+pixels = neopixel.NeoPixel(board.A1, NUM_PIXELS, brightness=0.1, auto_write=False)
+pixels.fill(0)
+pixels.show()
+
+mic = audiobusio.PDMIn(board.MICROPHONE_CLOCK, board.MICROPHONE_DATA,
+                       sample_rate=16000, bit_depth=16)
+
+# Record an initial sample to calibrate. Assume it's quiet when we start.
+samples = array.array('H', [0] * NUM_SAMPLES)
+mic.record(samples, len(samples))
+# Set lowest level to expect, plus a little.
+input_floor = normalized_rms(samples) + 10
+# Corresponds to sensitivity: lower means more pixels light up with lower sound
+input_ceiling = input_floor + 500
+peak = 0
+
+def wheel(wheel_pos):
+    # Input a value 0 to 255 to get a color value.
+    # The colours are a transition r - g - b - back to r.
+    if wheel_pos < 0 or wheel_pos > 255:
+        r = g = b = 0
+    elif wheel_pos < 85:
+        r = int(wheel_pos * 3)
+        g = int(255 - wheel_pos*3)
+        b = 0
+    elif wheel_pos < 170:
+        wheel_pos -= 85
+        r = int(255 - wheel_pos*3)
+        g = 0
+        b = int(wheel_pos*3)
+    else:
+        wheel_pos -= 170
+        r = 0
+        g = int(wheel_pos*3)
+        b = int(255 - wheel_pos*3)
+    return (r, g, b)
+
+def rainbow_cycle(wait):
+    for j in range(255):
+        for i in range(NUM_PIXELS):
+            pixel_index = (i * 256 // NUM_PIXELS) + j
+            pixels[i] = wheel(pixel_index & 255)
+        pixels.show()
+        time.sleep(wait)
+
+def audio_meter():
+    mic.record(samples, len(samples))
+    magnitude = normalized_rms(samples)
+    global peak
+
+    # Compute scaled logarithmic reading in the range 0 to NUM_PIXELS
+    c = log_scale(constrain(magnitude, input_floor, input_ceiling),
+                  input_floor, input_ceiling, 0, NUM_PIXELS)
+
+    # Light up pixels that are below the scaled and interpolated magnitude.
+    pixels.fill(0)
+    for i in range(NUM_PIXELS):
+        if i < c:
+            pixels[i] = volume_color(i)
+        # Light up the peak pixel and animate it slowly dropping.
+        if c >= peak:
+            peak = min(c, NUM_PIXELS - 1)
+        elif peak > 0:
+            peak = peak - 1
+        if peak > 0:
+            pixels[int(peak)] = PEAK_COLOR
+    pixels.show()
+
+pos = 0  # position
+direction = 1  # direction of "eye"
+
+def larsen_set(index, color):
+    if index < 0:
+        return
+    else:
+        pixels[index] = color
+
+def larsen(wait):
+    global pos
+    global direction
```

Comment:
Here, pass `pos` and `direction` in, and then return them as a tuple. You can do `return (pos, direction)`, and then in the call, do
```
pos, direction = larsen(wait, pos, direction)
```

Reply:
Making this change resulted in erratic animation speed – speeding up and slowing down. I left the function mostly unchanged for now.

Reply:
ok, not sure why, but good enuf